src/uct/ib/rdmacm/rdmacm_iface.c

```diff
@@ -281,39 +360,66 @@ static int uct_rdmacm_iface_process_event(uct_rdmacm_iface_t *iface, struct rdma
         /* Server - handle a connection request from the client */
         ucs_assert(iface->is_server);
         uct_rdmacm_iface_process_conn_req(iface, event, remote_addr);
+        ret_flags &= ~UCT_RDMACM_PROCESS_EVENT_ACK_EVENT_FLAG;
         break;
 
     case RDMA_CM_EVENT_REJECTED:
         /* Client - server rejected the connection request */
-        ucs_warn("rdmacm connection request to %s rejected",
-                  ucs_sockaddr_str(remote_addr, ip_port_str, UCS_SOCKADDR_STRING_LEN));
+        ucs_warn("rdmacm connection request to %s rejected, id %p",
+                  ucs_sockaddr_str(remote_addr, ip_port_str,
+                                   UCS_SOCKADDR_STRING_LEN), event->id);
 
-        destroy_cm_id = 1;
-        uct_rdmacm_client_handle_failure(iface, ep, UCS_ERR_CANCELED);
+        ret_flags |= UCT_RDMACM_PROCESS_EVENT_DESTROY_CM_ID_FLAG;
+        uct_rdmacm_client_handle_failure(iface, ep, UCS_ERR_REJECTED);
         break;
 
     case RDMA_CM_EVENT_ESTABLISHED:
         /* Client - connection is ready */
-        destroy_cm_id = 1;
+        ucs_assert(!iface->is_server);
+        ret_flags |= UCT_RDMACM_PROCESS_EVENT_DESTROY_CM_ID_FLAG;
+        if (ep) {
+            ep->status = UCS_OK;
+            ucs_queue_head_init(&completed_ops);
+            ucs_queue_for_each_safe(op, iter, &ep->ops, queue_elem) {
```

Comment:
why need to move items to another queue? can't we use ucs_queue_for_each_extract()?
we should NOT keep items with count>0 on the queue, they are meant to be eventually completed from another context

Reply:
`ucs_queue_for_each_safe` allows remove elements from the queue but don't free them.

Reply:
`ucs_queue_for_each_extract` stops iterating, that's not what's needed here

Reply:
we can use "1" to never stop, in any case we could then reuse the code for error handling

Reply:
Some operations can have `user_comp->count` > 1, the error handling code have to reset count to 0, flush doesn't

Reply:
the error handling in UCT should not reset comp to 0, only UCP can do that, there could be double free. see uct_rc_txqp_purge_outstanding()

Reply:
Sorry for misleading, error handling can extract all entries (ucs_queue_for_each_extract with always true condition), flush can't

Reply:
why not? after connection is established all comps should be invoked no?

Reply:
even if user requested uct completion count = 10?

Reply:
as discussed, need to fix comp counting, thanks for catching this!